### PR TITLE
[WIP] Adds new options to allow auto ignoring sounds and hostile creatures based on safe mode

### LINKED ISSUE #28672: Extremely annoying (I)gnore, and how it could be fixed
**Is your feature request related to a problem? Please describe.**  
While (for example) moving 100 two-by-fours from one space to another using the advanced inventory, if there's any kind of loud noise, player is stuck having to hold the I key to ignore the sounds. I'm pretty sure everyone is already familiar with and have encountered the problem at least one time in the game.

**Describe the solution you'd like**  
IMHO, we should have something like "ignore sounds entirely" option, or we should use the safe mode to say that we don't want safety while moving these items. There is a problem though, and that's: What happens if a zombie comes and kills me while i'm moving the stuff and have the safe mode off?
Well, I think safe mode works as a "sight safe mode" currently, so adding sound to that would be weird and hard. Maybe it's better to add an additional safe mode called "sound safe mode" which controls safety for sound only. Also if you decide to add scent detection to the player some time in the future (Like, for example, "you can smell rotten meat here!"), you could add another safe mode for scents called "scent safe mode" which would control the scent detection only. With more safe modes, every player has more in-game control in my opinion.

**Additional context**  
For me, this is a must-do "fix". I've seen people getting really mad about having to press 'I' to ignore all the messages, and i've experienced it myself too. I think it's super annoying, and with more safe modes and more control, this could be fixed nicely.

Thank you for your time!

### LINKED ISSUE #36151: Make interrupt messages being an option to turn on/off
**Is your feature request related to a problem? Please describe.**  
A problem, well, when you have auto-foraging turned on, and you're getting 

https://media.discordapp.net/attachments/252368987210514433/656078283619958784/aaa.png
https://media.discordapp.net/attachments/252368987210514433/656078291903447051/bbb.png

every step you make, you tell me. If I may be so blunt: It's fucking annoying! I need it to go away! I like living on the edge! If I die, it's my fault. Not the games. No handholding for me.

Also same stuff when you have 1-shot weapons. You reload, message "blabla close", you aim, message "blabla close", you shot, you reload, message, you aim, message, you get the picture. That's not my definition of fun. BEing interrupted at every other turn.

Please make these interrupt messages an option to be turned off / on by the user! "Discussion"/Feedback going here: https://discordapp.com/channels/252365627543519232/252368987210514433/656078284164956170

### PR DESCRIPTION
**Pull request by Ramza13**
#### Summary
SUMMARY: Interface "Adds new options to allow auto ignoring sounds and hostile creatures."

#### Purpose of change
Fixes #28672
Fixes #36151
Expanded and reworked version of #35436
Adds new options to allow auto ignoring based on safe mode and distance.  

#### Describe the solution
Added new settings.  Auto ignore sound/hostile mode can be set to ignore always, never or when safe 
mode is on or off.  There is a distance setting so you can choose to only ignore sounds or creatures past a specific distance away.  No matter what settings are chosen dangerously close monster warnings will still appear as will creatures coming out of portals

#### Describe alternatives you've considered
Various other combos of settings.  A separate distance setting for safe mode on vs. off

### CODE REVIEW

**Comment by ave-** [comment]
Is z-level taken into account? I do not want to be warned about a sound 5 z-levels below me if I have a setting to warn about sound in 5-tile radius. There should be either a separate setting for z-level or different z-level sounds should be ignored entirely.

**Comment by Ramza13** [comment]
> 
> 
> Is z-level taken into account? I do not want to be warned about a sound 5 z-levels below me if I have a setting to warn about sound in 5-tile radius. There should be either a separate setting for z-level or different z-level sounds should be ignored entirely.

Presently z levels are counted in the distance.  However each z level counts as 10 squares so in your 5  square example all sounds on different z levels would be ignored.

File: src/options.cpp [hunk 1]
```diff
@@ -1174,6 +1174,34 @@ void options_manager::add_options_general()
 
     mOptionsSort["general"]++;
 
+    add( "AUTOIGNORESOUND", "general", translate_marker( "Auto ignore sound mode" ),
+         translate_marker( "Always: All sounds auto ignore distance away and further will be ignored.  SafeOn: All sounds auto ignore distance away and further will be ignored while safe mode is on.  SafeOff: All sounds auto ignore distance and further away will be ignored while safe mode is off.  None: No sounds will be auto ignored" ),
```
Comment (Night-Pryanik) [inline src/options.cpp]:
```suggestion
         translate_marker( "Always: All sounds auto ignore distance away and further will be ignored.  SafeOn: All sounds auto ignore distance away and further will be ignored while safe mode is on.  SafeOff: All sounds auto ignore distance and further away will be ignored while safe mode is off.  None: No sounds will be auto ignored." ),
```

File: src/options.cpp [hunk 2]
```diff
@@ -1174,6 +1174,34 @@ void options_manager::add_options_general()
...
+    { { "ALWAYS", translate_marker( "Always" ) }, { "SAFEON", translate_marker( "SafeOn" ) }, { "SAFEOFF", translate_marker( "SafeOff" ) }, { "NONE", translate_marker( "None" ) } },
+    "NONE"
+       );
+
+    add( "AUTOIGNORESOUNDDISTANCE", "general", translate_marker( "Auto ignore sound distance" ),
+         translate_marker( "Only sounds this distance and further away will be auto ignored" ),
```
Comment (Night-Pryanik) [inline src/options.cpp]:
```suggestion
         translate_marker( "Only sounds this distance and further away will be auto ignored." ),
```

File: src/options.cpp [hunk 3]
```diff
@@ -1174,6 +1174,34 @@ void options_manager::add_options_general()
...
+         0, 30, 0
+       );
+
+    get_option( "AUTOIGNORESOUNDDISTANCE" ).setPrerequisites( "AUTOIGNORESOUND", { "ALWAYS", "SAFEON", "SAFEOFF" } );
+
+    add( "AUTOIGNOREHOSTILE", "general", translate_marker( "Auto ignore hostile mode" ),
+         translate_marker( "No matter what is chosen dangerously close enemies will not be ignored. All: All hostiles auto ignore distance and further away will be ignored.  SafeOn: All hostiles auto ignore distance and further away will be ignored while safe mode is on.  SafeOff: All hostiles auto ignore distance and further away will be ignored while safe mode is off.  None: No hostiles will be auto ignored" ),
```
Comment (Night-Pryanik) [inline src/options.cpp]:
```suggestion
         translate_marker( "No matter what is chosen dangerously close enemies will not be ignored. All: All hostiles auto ignore distance and further away will be ignored.  SafeOn: All hostiles auto ignore distance and further away will be ignored while safe mode is on.  SafeOff: All hostiles auto ignore distance and further away will be ignored while safe mode is off.  None: No hostiles will be auto ignored." ),
```

File: src/options.cpp [hunk 4]
```diff
@@ -1174,6 +1174,34 @@ void options_manager::add_options_general()
...
+    { { "ALWAYS", translate_marker( "Always" ) }, { "SAFEON", translate_marker( "SafeOn" ) }, { "SAFEOFF", translate_marker( "SafeOff" ) }, { "NONE", translate_marker( "None" ) } },
+    "NONE"
+       );
+
+    add( "AUTOIGNOREHOSTILEDISTANCE", "general", translate_marker( "Auto ignore hostile distance" ),
+         translate_marker( "Only hostiles this distance and further away will be auto ignored" ),
```
Comment (Night-Pryanik) [inline src/options.cpp]:
```suggestion
         translate_marker( "Only hostiles this distance and further away will be auto ignored." ),
```

**Comment by Ramza13** [comment]
Added in an option to allow auto ignoring of sounds on different z levels regardless of distance

**Comment by kevingranade** [comment]
This adds way too many new options, we already have too many, and we need really strong justification for new ones.

**Comment by Ramza13** [comment]
> 
> 
> This adds way too many new options, we already have too many, and we need really strong justification for new ones.

Ok fair enough.  If I just made it two options one for when to auto-ignore sounds/hostiles and a second for distance would that be better or would you prefer a non option based solution?

**Comment by Ramza13** [comment]
Trimmed this back to 2 options.  Now there is a single setting for auto ignoring both sounds and hostiles and a single distance.  Ignoring z levels is also gone but each one is 10 horizontal distance so its not so bad.

File: src/options.cpp [hunk 5]
```diff
@@ -1174,6 +1174,21 @@ void options_manager::add_options_general()
 
     mOptionsSort["general"]++;
 
+    add( "AUTOIGNOREMODE", "general", translate_marker( "Auto ignore sounds/hostiles mode" ),
+         translate_marker( "While auto ignore is active you will not be interrupted by sounds or spotted hostiles. Always: Auto ignore will always be active.  SafeOn: Auto ignore will be active while safe mode is on.  SafeOff: Auto ignore will be active while safe mode is off.  Never: Auto ignore will never be active." ),
```
Comment (AMurkin) [inline src/options.cpp]:
```suggestion
         translate_marker( "While auto ignore is active you will not be interrupted by sounds or spotted hostiles.  Always: Auto ignore will always be active.  SafeOn: Auto ignore will be active while safe mode is on.  SafeOff: Auto ignore will be active while safe mode is off.  Never: Auto ignore will never be active." ),
```

File: src/options.cpp [hunk 6]
```diff
@@ -1174,6 +1174,21 @@ void options_manager::add_options_general()
...
+    { { "ALWAYS", translate_marker( "Always" ) }, { "SAFEON", translate_marker( "SafeOn" ) }, { "SAFEOFF", translate_marker( "SafeOff" ) }, { "NEVER", translate_marker( "Never" ) } },
+    "NEVER"
+       );
+
+    add( "AUTOIGNOREDISTANCE", "general", translate_marker( "Auto ignore starting distance" ),
+         translate_marker( "Only sounds/hostiles this distance and further away will be auto ignored.  Each Z Level counts as 10 horizontal distance." ),
```
Comment (AMurkin) [inline src/options.cpp]:
```suggestion
         translate_marker( "Only sounds/hostiles this distance and further away will be auto ignored.  Each z-Level counts as 10 horizontal distance." ),
```

**Comment by Balgias** [comment]
Anyone who downvoted this because you love the warning spam in this game I advise you fuck off forever throw your PC out a window and go eat trash.

**Pull request closed without merging**